crates/shielded_token/src/masp/shielded_sync/dispatcher.rs

```diff
@@ -381,7 +382,101 @@ where
+            self.client.capabilities().needs_witness_map_update();
+        while let Some(first_tx) = indexed_txs.next() {
+            // Take only Transactions belonging to one block
+            let mut block_txs = vec![first_tx];
+            while let Some(next_tx) = indexed_txs
+                .next_if(|next_tx| next_tx.0.height == block_txs[0].0.height)
+            {
+                block_txs.push(next_tx);
+            }
+            // Do not apply these Transactions if they have been applied before
+            if !needs_witness_map_update
+                || Some(&block_txs[0].0) <= last_witnessed_tx.as_ref()
+            {
+                continue;
+            }
+            // Save the initial state of the ShieldedContext in case reversion
+            // to initial state is necessary
+            let initial_note_index = self.ctx.note_index.clone();
+            let initial_witness_map = self.ctx.witness_map.clone();
+            let initial_tree = self.ctx.tree.clone();
+            let mut anchor_exists = false;
+            // Valid Transaction orderings are always the concatenation of two
+            // subsequences of the ordered events. Let's iterate through
+            // possible values of the first subsequence
+            for subset in block_txs.iter().powerset() {
+                // First make sure that we start with a clean state uncorrupted
+                // by previous trials
+                self.ctx.note_index = initial_note_index.clone();
+                self.ctx.witness_map = initial_witness_map.clone();
+                self.ctx.tree = initial_tree.clone();
+                let mut fee_transfers = BTreeSet::new();
+                // Apply the first subsequence, which corresponds to the
+                // Transactions that were first applied for the fees
+                for (indexed_tx, stx_batch) in &subset {
+                    tracing::debug!("Transaction Index: {:?}", indexed_tx);
+                    fee_transfers.insert(indexed_tx);
+                    self.ctx.update_witness_map(*indexed_tx, stx_batch).await?;
+                }
+                // Apply the second subsequence, which corresponds to the
+                // Transactions not used in the fees
+                for (indexed_tx, stx_batch) in &block_txs {
+                    // The second subsequence consists of elements not in the
+                    // first
+                    if !fee_transfers.contains(indexed_tx) {
+                        tracing::debug!("Transaction Index: {:?}", indexed_tx);
+                        self.ctx
+                            .update_witness_map(*indexed_tx, stx_batch)
+                            .await?;
+                    }
+                }
+                // Compute what the note commitment tree root would look like
+                // after applying the above order and check if it's recognized
+                // by the protocol
+                let root = self.ctx.tree.root();
```

Comment:
for this search, we only need to update the tree, not the witness maps and note indexes. Once we have found the order, we can update those. This should save a lot of computation and cloning here.